File: ios/Models/CBReactNativeError.swift
//
//  CBError+Extensions.swift
//  ChargebeeReactNative
//
//  Created by cb-haripriyan on 20/02/23.
//

import Foundation
import Chargebee

enum CBReactNativeError: Int, Error {
    
    case unknown = 0
    
    // MARK: Chargebee Errors
    case invalidSdkConfiguration = 1000
    case invalidCatalogVersion = 1001
    case cannotMakePayments = 1002
    case noProductToRestore = 1003
    
    // MARK: Store Errors
    case invalidOffer = 2001
    case invalidPurchase = 2002
    case invalidSandbox = 2003
    case networkError = 2004
    case paymentFailed = 2005
    case paymentNotAllowed = 2006
    case productNotAvailable = 2007
    case purchaseNotAllowed = 2008
    case purchaseCancelled = 2009
    case storeProblem = 2010
    case invalidReceipt = 2011
    case requestFailed = 2012
    case productPurchasedAlready = 2013
    
    // MARK: Restore Error
    case noReceipt = 2014
    case refreshReceiptFailed = 2015
    case restoreFailed = 2016
    
    // MARK: General Errors
    case systemError = 3000
    
}

extension CBReactNativeError {
    static func errorCode(purchaseError: CBPurchaseError) -> CBReactNativeError {
        switch purchaseError {
        case .productIDNotFound, .productsNotFound, .productNotAvailable:
            return CBReactNativeError.productNotAvailable
        case .skRequestFailed:
            return CBReactNativeError.requestFailed
        case .cannotMakePayments:
            return CBReactNativeError.cannotMakePayments
        case .noProductToRestore:
            return CBReactNativeError.noProductToRestore
        case .invalidSDKKey:
            return CBReactNativeError.invalidSdkConfiguration
        case .invalidCustomerId:
            return CBReactNativeError.invalidSdkConfiguration
        case .invalidCatalogVersion:
            return CBReactNativeError.invalidCatalogVersion
        case .userCancelled:
            return CBReactNativeError.purchaseCancelled
        case .paymentFailed:
            return CBReactNativeError.paymentFailed
        case .invalidPurchase:
            return CBReactNativeError.invalidPurchase
        case .invalidClient, .privacyAcknowledgementRequired:
            return CBReactNativeError.purchaseNotAllowed
        case .networkConnectionFailed:
            return CBReactNativeError.networkError
        case .unknown:
            return CBReactNativeError.unknown
        case .paymentNotAllowed:
            return CBReactNativeError.paymentNotAllowed
        case .invalidOffer, .invalidPrice, .invalidPromoCode, .invalidPromoOffer:
            return CBReactNativeError.invalidOffer
        case .invalidSandbox:
            return CBReactNativeError.invalidSandbox
        }
    }
    

}

File: android/src/main/java/com/chargebee/android/reactnative/models/CBReactNativeError.kt
package com.chargebee.android.reactnative.models

import com.android.billingclient.api.BillingClient
import com.android.billingclient.api.BillingClient.BillingResponseCode.*

enum class CBReactNativeError(val code: Int) {

  UNKNOWN(0),

  // Chargebee Errors
  INVALID_SDK_CONFIGURATION(1000),
  INVALID_CATALOG_VERSION(1001),
  CANNOT_MAKE_PAYMENTS(1002),
  NO_PRODUCT_TO_RESTORE(1003),
  INVALID_RESOURCE(1004),

  // Store Errors
  INVALID_OFFER(2001),
  INVALID_PURCHASE(2002),
  INVALID_SANDBOX(2003),
  NETWORK_ERROR(2004),
  PAYMENT_FAILED(2005),
  PAYMENT_NOT_ALLOWED(2006),
  PRODUCT_NOT_AVAILABLE(2007),
  PURCHASE_NOT_ALLOWED(2008),
  PURCHASE_CANCELLED(2009),
  STORE_PROBLEM(2010),
  INVALID_RECEIPT(2011),
  REQUEST_FAILED(2012),
  PRODUCT_PURCHASED_ALREADY(2013),

  // Restore Error
  NO_RECEIPT(2014),
  REFRESH_RECEIPT_FAILED(2015),
  RESTORE_FAILED(2016),

  // General Errors
  SYSTEM_ERROR(3000);

  companion object {
    fun fromBillingCode(code: Int) : CBReactNativeError {
      return when (code) {
        FEATURE_NOT_SUPPORTED, BILLING_UNAVAILABLE, ITEM_NOT_OWNED -> PURCHASE_NOT_ALLOWED
        ERROR, SERVICE_UNAVAILABLE, SERVICE_DISCONNECTED, SERVICE_TIMEOUT -> STORE_PROBLEM
        USER_CANCELED -> PURCHASE_CANCELLED
        ITEM_UNAVAILABLE -> PRODUCT_NOT_AVAILABLE
        DEVELOPER_ERROR -> INVALID_PURCHASE
        ITEM_ALREADY_OWNED -> PRODUCT_PURCHASED_ALREADY
        else -> UNKNOWN
      }
    }
  }
}
